theseus/optimizer/nonlinear/nonlinear_least_squares.py

```diff
@@ -4,18 +4,58 @@
 # LICENSE file in the root directory of this source tree.
 
 import abc
-from typing import Any, Dict, Optional, Type
+import warnings
+from typing import Any, Callable, Dict, NoReturn, Optional, Tuple, Type, Union
 
 import torch
 
 from theseus.core import Objective
-from theseus.optimizer import Linearization
-from theseus.optimizer.linear import LinearSolver
-from theseus.optimizer.linear.dense_solver import CholeskyDenseSolver
+from theseus.optimizer import Linearization, OptimizerInfo
+from theseus.optimizer.linear import (
+    LinearSolver,
+    CholeskyDenseSolver,
+    LUCudaSparseSolver,
+)
+from .nonlinear_optimizer import (
+    BackwardMode,
+    NonlinearOptimizer,
+    NonlinearOptimizerInfo,
+    NonlinearOptimizerStatus,
+)
 
-from .nonlinear_optimizer import NonlinearOptimizer
 
+EndIterCallbackType = Callable[
+    ["NonlinearOptimizer", NonlinearOptimizerInfo, torch.Tensor, int], NoReturn
+]
 
+
+# Base class for all optimizers for NLLS problems,
+# providing the skeleton of the
+# optimization loop. Subclasses need to implement the following method:
+#
+#   - `compute_delta`: returns a descent direction given the current values
```

Comment:
This (or a similar fucntion) should be in trust region?

Reply:
In response to Mustafa, I think line search is a more general procedure, and I'd argue it would even belong in `NonlinearOptimizer`. Also, a method is probably not the right abstraction, but something like a member with type specified at construction time, to specify which line search method to use, each implemented in a different class. 

But in any case, I'd prefer not to include anything for `line_search()` until we have a concrete implementation of a differentiable line search method with batching support, and we know what an implementation would look like.